test_runner/src/main/kotlin/ftl/shard/Shard.kt

```diff
@@ -100,68 +111,87 @@ object Shard {
-        val testCount =
-            if (testCases.isEmpty()) 0
-            else testCases.filter { it.time > 0.0 }.takeIf { it.isNotEmpty() }?.size ?: 1
+        val testCount = getNumberOfNotIgnoredTestCases(testCases)
 
         // If maxShards is infinite or we have more shards than tests, let's match it
-        val shardsCount = if (maxShards == -1 || maxShards > testCount) testCount else maxShards
+        val shardsCount = matchNumberOfShardsWithTestCount(maxShards, testCount)
 
         // Create the list of shards we will return
-        if (shardsCount <= 0) {
-            val platform = if (args is IosArgs) "ios" else "android"
-            throw FlankFatalError(
-                """Invalid shard count. To debug try: flank $platform run --dump-shards
+        assertNotTrue(shardsCount <= 0) {
+            FlankFatalError(
+                """Invalid shard count. To debug try: flank ${args.platformName} run --dump-shards
                     | args.maxTestShards: ${args.maxTestShards}
                     | forcedShardCount: $forcedShardCount 
                     | testCount: $testCount 
                     | maxShards: $maxShards 
                     | shardsCount: $shardsCount""".trimMargin()
             )
         }
-        var shards = List(shardsCount) { TestShard(0.0, mutableListOf()) }
+        var shards = createListOfShards(shardsCount)
 
         testCases.forEach { testMethod ->
-            val shard = shards.first()
-
-            shard.testMethods.add(testMethod)
-            shard.time += testMethod.time
-
-            // Sort the shards to keep the most empty shard first
-            shards = shards.sortedBy { it.time }
+            addTestMethodToMostEmptyShard(shards, testMethod)
+            shards = shards.mostEmptyFirst()
         }
 
-        val allTests = testsToRun.size // zero when test targets is empty
-        val cacheHit = allTests - cacheMiss
-        val cachePercent = if (allTests == 0) 0.0 else cacheHit.toDouble() / allTests * 100.0
+        val cacheTestCounter = CacheTestCounter(testsToRun.size, cacheMiss)
         println()
-        println("  Smart Flank cache hit: ${cachePercent.roundToInt()}% ($cacheHit / $allTests)")
+        println("  Smart Flank cache hit: ${cacheTestCounter.cachePercent.roundToInt()}% (${cacheTestCounter.cacheHit} / ${cacheTestCounter.allTestCount})")
         println("  Shard times: " + shards.joinToString(", ") { "${it.time.roundToInt()}s" } + "\n")
 
         return shards
     }
 
+    private fun addTestMethodToMostEmptyShard(shards: List<TestShard>, testMethod: TestMethod) {
+        val mostEmptyShard = shards.first()
+        mostEmptyShard.testMethods.add(testMethod)
+        mostEmptyShard.time += testMethod.time
+    }
+
+    private val IArgs.platformName
+        get() = if (this is IosArgs) "ios" else "android"
+
+    private fun assertNotTrue(condition: Boolean, lazyException: () -> Throwable) {
```

Comment:
I am not sure if this abstraction is really needed.

Reply:
I will remove this